### ISSUE TITLE
How do I negate a bool?

### ISSUE THREAD

**Issue by jyn514** [CraneStation/cranelift#922]
I'm trying to compile this simple C program:

```c
int main() {
    return !1;
}
```

I'll spare you the parsing, what my program does after implicit casts is:

- create a literal using `let val = builder.ins().iconst(types::I32, 1)`, where `builder` is a `FunctionBuilder;`
- cast 1 to `B1` using `let val = builder.ins().icmp_imm(condcodes::IntCC::NotEqual, val, 0);`
- negate the `B1` using `let val = builder.ins().bnot(val);`
- extend the `B1` to `I32` using `builder.ins().bint(types::I32, val)`

This gives me a verifier error: `VerifierError { location: inst2, message: "v2 is a real GPR value defined by a ghost instruction" }`.

What, if anything, am I doing wrong? I see that `rustc_codegen_cranelift` [extends the bool to an I32](https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/src/base.rs#L291), do I need to do that?

**Comment by jyn514** [comment]
Cranelift seems to think `bnot(val)` where val is a B1 [is an illegal instruction](https://docs.rs/cranelift-codegen/0.36.0/src/cranelift_codegen/verifier/liveness.rs.html#80), what should I use instead?

**Comment by bnjbvr** [comment]
Thanks for opening an issue! `bnot` is indeed only encoded for i32 and i64 on x86; you could use `bint` to transform the B1 value into an I32 or an I64, and then use `bnot` on the result.

Alternatively, we could also have all the boolean operations get encodings on x86, since they're well-typed. Or better, legalization could implicitly do a conversion to the equivalent int type, apply the requested operation and then convert back to an bool type.

**Comment by bjorn3** [comment]
`bnot` on `0i32` would give `i32::max_value()` which is not a valid bool, only 0 and 1 are. Thats why I did `bint` + `icmp_imm eq val, 0`.

**Comment by bnjbvr** [comment]
Yeah, I'd expect there be an instruction to convert back from int to bool in a dynamic way; not sure it exists, though.

**Comment by jyn514** [comment]
I think converting to int and back is a good workaround for now. Encoding boolean operations would be ideal but I can live without that.

Where can I find what instructions are valid for which architectures? Currently I'm doing a lot of guess and check and it would be nice to have a reference. The docs don't mention any limitations on the type, see for example [`bnot`](https://docs.rs/cranelift-codegen/0.36.0/cranelift_codegen/ir/trait.InstBuilder.html#method.bnot) which just says 'Bitwise not.'

**Comment by bjorn3** [comment]
> edit: `icmp_imm ne x, 0` as noted by @sunfishcode on IRC.

For `bnot` it is more efficient to do the not and convert to bool in one step using `icmp_imm eq x, 0`.

> Where can I find what instructions are valid for which architectures?

All instructions without a prefix like `x86_` should work on all isas eventually, however many dont have an encoding/legalization yet.

There is no list of allowed pairs between instruction and isa at the moment. You can try them yourself or read the source.

**Comment by jyn514** [comment]
> For bnot it is more efficient to do the not and convert to bool in one step using icmp_imm eq x, 0.

I ended up implementing it that way for ints, but I still need to deal with bools for things like `!(a < b && b < c)`.

> There is no list of allowed pairs between instruction and isa at the moment. You can try them yourself or read the source.

That's a shame... Someone mentioned an IRC channel, how would I find that if I had more questions? I don't want to keep opening github issues for things that aren't really bugs.

**Comment by bjorn3** [comment]
> Someone mentioned an IRC channel, how would I find that if I had more questions?

#cranelift on irc.mozilla.org
